
<div data-component="Callout" data-prop="W3sidHlwZSI6Im1keEF0dHJpYnV0ZSIsIm5hbWUiOiJ2YXJpYW50IiwidmFsdWUiOiJpbXBvcnRhbnQifV0=">
  <p>The end-of-life date for this agent version is July 29, 2019. To update to the latest agent version, see <a href="/docs/agents/manage-apm-agents/installation/update-new-relic-agent">Update the agent</a>. For more information, see <a href="/docs/agents/manage-apm-agents/maintenance/new-relic-agent-plugin-end-life-policy">End-of-life policy</a>.</p>
</div>
<h3>Notes</h3>
<p>This release of the Python agent contains support for web applications and worker processes using 'gearman' for background task execution. The agent also switches to using the 'json' package from the standard library instead of a separate 'simplejson' package. This should result in potential reductions in base level memory usage by the agent and a reduction of CPU overhead for some environments.</p>
<p>The agent can be installed using easy_install/pip/distribute via the <a href="https://pypi.python.org/pypi/newrelic">Python Package Index</a> or can be downloaded directly from our <a href="http://download.newrelic.com/python_agent/release/">download site</a>.</p>
<p>For a list of known issues with the Python agent see our online help article on the <a href="https://newrelic.com/docs/python/status-of-python-agent">status of the Python agent</a>.</p>
<h3>New Features:</h3>
<ul>
  <li>Instrumentation now provided for the <a href="https://pypi.python.org/pypi/gearman/">gearman</a> package. The support covers use of both the gearman client and worker interfaces. In the case of workers, each task executed by the worker will be recorded as a background task and be displayed as such in the New Relic UI.</li>
  <li>The <code>ignore_errors</code> argument to <code>newrelic.agent.record_exception()</code> can now be a callable in addition to being able to pass a sequence. When it is a callable, the callback will be called with the three arguments <code>exc</code>, <code>value</code> and <code>tb</code>, being the same values as returned by <code>sys.exc_info()</code>. The callback should return <code>True</code> if the exception is to be ignored. <code>False</code> if the exception should never be ignored regardless of any other checks, and <code>None</code> if subsequent checks and inbuilt rules should determine if the exception should be ignored. A callback would normally return either <code>True</code> or <code>None</code>.</li>
</ul>
<h3>Bug fixes/Improvements:</h3>
<ul>
  <li>The bundled 'simplejson' package has now been removed and is no longer used. This was previously used due to the requirement to support Python 2.5, for which support was removed in version 2.0.0 of the agent. Instead of simplejson, the 'json' package contained in the Python standard library is now used instead. This should see the base level memory usage of the agent drop as a result. Further, in environments where a C compiler was not available and the C extension modules could not be compiled, there should now be a minimal reduction in the CPU overhead when the agent is uploading data to our back end. This is because the optimised C extension in the json package will always be available as part of the Python standard library and will always be able to be used.</li>
  <li>If a WSGI application was returning an iterable such as a generator, and an exception was raised when a specific part of the response content was yielded from the generator, the details of the exceptions were not being recorded.</li>
  <li>In a multithreaded web application where deferred module imports were being performed in secondary threads, agent registration could fail the first time due to concurrent changes made by the secondary threads to <code>sys.modules</code>. Registration would succeed on a subsequent attempt. This bug was only occurring in Python 2 and was introduced in version 2.0.0 of the agent when Python 3 support was added.</li>
  <li>Custom parameters which were explicitly supplied to <code>newrelic.agent.record_exception()</code> were being ignored and were not appearing in the error details page in the UI. Instead, only custom parameters added using <code>newrelic.agent.add_custom_parameters()</code> against the web transaction or background task itself were being shown. This bug was introduced in version 1.13.0 of the agent when support for cross application tracing was added.</li>
  <li>The port number used in the URL for a web external call, was not being retained when the URL was added as a parameter against the web external node for transaction traces. The port would therefore not be displayed when drilling down into the details of a web external node in a transaction trace.</li>
  <li>The <code>newrelic-admin validate-config</code> command will now work for enterprise high security mode accounts, provided of course that no setting is otherwise specified in the agent configuration file which is in conflict with that mode. That is, features such as SSL would still need to be enabled by the <code>ssl</code> setting and the <code>capture_params</code> setting indicating whether URL query string parameters should be captured, also set to <code>false</code>.</li>
  <li>When using CherryPy, in addition to the explicit exception type 'NotFound' being ignored as an error, raising of a 'HTTPError' exception where the status is 404 will also be ignored. Similarly, instances of the 'HTTPError' exception will be ignored as an error when they are for a HTTP error in the 30X range.</li>
</ul>
